### PR TITLE
Improve Azure deployment reconciler and add tests

### PR DESCRIPTION
**Pull request by matthchr**
**What this PR does / why we need it**:
- Fixes a number of issues with the Azure deployment reconciler.
- Adds various edge-case tests to ensure that scenarios such as creating 
  a resource before it's owner is created, or updating a failed resource, work
  correctly.

**How does this PR make you feel**:
gif

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains tests

### CODE REVIEW

**Comment by matthchr** [comment]
Ignore the first commit (based on my VMSS branch, will go away once that merges).

This is a draft PR just so folks are aware of the upcoming tests and changes. I think after this we should have enough test coverage to be more confident about the reconciler behavior. 

Thinking we can follow this PR up with some refactoring around:
1. Resource states
2. Code structure in the reconciler

File: hack/generated/pkg/testcommon/kube_per_test_context.go [hunk 1]
```diff
@@ -191,7 +197,7 @@ func (tc KubePerTestContext) CreateNewTestResourceGroup(wait WaitCondition) (*re
 		// doesn't stop polling resources in "Deleting" state and so when running recordings
 		// different runs end up polling different numbers of times. Ensuring we reach a state
 		// the controller deems terminal (Deleted) resolves this issue.
-		tc.G.Eventually(rg, 2*time.Minute).Should(tc.Match.BeDeleted())
+		tc.G.Eventually(rg, 5*time.Minute).Should(tc.Match.BeDeleted())
```
Comment (Porges) [inline hack/generated/pkg/testcommon/kube_per_test_context.go:200]: Perhaps should be in a set of “test constants” somewhere

File: hack/generated/pkg/reconcilers/deployment_error_classifier_test.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"testing"
+
+	"github.com/Azure/azure-service-operator/hack/generated/pkg/armclient"
+	"github.com/Azure/azure-service-operator/hack/generated/pkg/reconcilers"
+	. "github.com/onsi/gomega"
+)
+
+var badRequestError = armclient.DeploymentError{
+	Code:    "BadRequest",
+	Message: "That was not a good request",
```
Comment (Porges) [inline hack/generated/pkg/reconcilers/deployment_error_classifier_test.go:18]: 😁

File: hack/generated/pkg/testcommon/kube_per_test_context.go (hunk 1 above)
Comment (matthchr) [reply hack/generated/pkg/testcommon/kube_per_test_context.go:200]: Yeah this is awkward because in replay mode it doesn't really need to be this long but in record mode I've seen a few tests fail because it was too short. Not consistently though. I figured bumping it up was _probably_ ok, but can pull it out into a constant.

File: hack/generated/controllers/deployment_reconciler_negative_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+	compute "github.com/Azure/azure-service-operator/hack/generated/_apis/microsoft.compute/v1alpha1api20201201"
+	resources "github.com/Azure/azure-service-operator/hack/generated/_apis/microsoft.resources/v1alpha1api20200601"
+	storage "github.com/Azure/azure-service-operator/hack/generated/_apis/microsoft.storage/v1alpha1api20210401"
+	"github.com/Azure/azure-service-operator/hack/generated/pkg/armclient"
+	"github.com/Azure/azure-service-operator/hack/generated/pkg/reconcilers"
+	"github.com/Azure/azure-service-operator/hack/generated/pkg/testcommon"
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (theunrepentantgeek) [inline hack/generated/controllers/deployment_reconciler_negative_test.go:19]: Are these imports in the right order? (Channeling my inner @babbageclunk)

File: hack/generated/controllers/deployment_reconciler_negative_test.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	tc.Expect(err).ToNot(HaveOccurred())
+	updated := &storage.StorageAccount{}
+	tc.GetResource(objectKey, updated)
+
+	tc.Expect(updated.Annotations).ToNot(HaveKey(reconcilers.ResourceErrorAnnotation))
+}
+
+// There are two ways that a deployment can fail. It can be rejected when initially
+// submitted to the Azure API, or it can be accepted and then report a failure during
+// long running operation polling. This ensures that the first case is handled correctly.
```
Comment (theunrepentantgeek) [inline hack/generated/controllers/deployment_reconciler_negative_test.go:127]: These comments on the intent of each test are really valuable, especially when there's a lot going on with the test; maybe add these to other tests in this file too?

File: hack/generated/pkg/reconcilers/deployment_error_classifier.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	if len(deploymentError.Details) == 0 {
+		// Default to retrying if we're asked to classify an error with no details
+		return DeploymentErrorRetryable
+	}
+
+	// Classify all of the details -- there may ALWAYS be only one but
+	// since the API technically allows a list just deal with it in case
+	// it actually happens in some rare case.
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/reconcilers/deployment_error_classifier.go:27]: 😁

File: hack/generated/pkg/reconcilers/deployment_error_classifier.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		"AuthorizationFailed",
+		"AllocationFailed",
+		"InvalidResourceReference",
+		"InvalidSubscriptionRegistrationState",
+		"LinkedAuthorizationFailed",
+		"MissingRegistrationForLocation",
+		"MissingSubscriptionRegistration",
+		"NoRegisteredProviderFound",
+		"NotFound",
+		// It sounds weird to retry on the below but according to the docs it's a quota issue, so
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/reconcilers/deployment_error_classifier.go:60]:
```suggestion
		// It sounds weird to retry on the "OperationNotAllowed" but according to the docs it's a quota issue, so
```

File: hack/generated/pkg/testcommon/ensure.go [hunk 7]
```diff
@@ -31,8 +31,8 @@ func NewEnsure(c client.Client, stateAnnotation string, errorAnnotation string)
 	}
 }
 
-// Provisioned checks to ensure the provisioning state of the resource is successful.
-func (e *Ensure) Provisioned(ctx context.Context, obj runtime.Object) (bool, error) {
+// State checks to ensure the provisioning state of the resource the target state.
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/testcommon/ensure.go:34]:
```suggestion
// State checks to check whether the provisioning state of the resource has the target state.
```

"Ensure" is an active word that implies this method will make any changes required for the resource to be in target state; I don't think that's what you mean.

Comment (theunrepentantgeek) [reply hack/generated/pkg/testcommon/ensure.go:34]: Maybe call the method `HasState()`?

**Review by theunrepentantgeek — APPROVED** [review]
Some minor suggestions, but otherwise looks good.

File: hack/generated/pkg/testcommon/kube_per_test_context.go (hunk 1 above)
Comment (matthchr) [reply hack/generated/pkg/testcommon/kube_per_test_context.go:200]: Pulled this up into a `ResourceGroupDeletionWaitTime` at the top of this file for now

File: hack/generated/controllers/deployment_reconciler_negative_test.go (hunk 3 above)
Comment (matthchr) [reply hack/generated/controllers/deployment_reconciler_negative_test.go:19]: Good eye, fixed. Not sure why Goland didn't flag that

File: hack/generated/controllers/deployment_reconciler_negative_test.go (hunk 4 above)
Comment (matthchr) [reply hack/generated/controllers/deployment_reconciler_negative_test.go:127]: Done

File: hack/generated/pkg/reconcilers/deployment_error_classifier.go (hunk 6 above)
Comment (matthchr) [reply hack/generated/pkg/reconcilers/deployment_error_classifier.go:60]: Fixed

File: hack/generated/pkg/testcommon/ensure.go (hunk 7 above)
Comment (matthchr) [reply hack/generated/pkg/testcommon/ensure.go:34]: > "Ensure" is an active word that implies this method will make any changes required for the resource to be in target state; I don't think that's what you mean.

The actual struct is already called `Ensure` - but the methods hanging off as you note do not actually make changes, they just check them. This same comment style is repeated for the other methods `Provisioned`, `Failed`, etc.

[Merriam-Webster](https://www.merriam-webster.com/dictionary/ensure) says:
> Ensure: to make sure, certain, or safe : guarantee

whereas [Cambridge dictionary](https://dictionary.cambridge.org/us/dictionary/english/ensure) agrees with you:
> To make something certain to happen

I'm going to hold off making a change on this - if we do want to change it we should change the whole class, not just this one place.

Comment (matthchr) [reply hack/generated/pkg/testcommon/ensure.go:34]: Changed to `HasState`

Comment (matthchr) [reply hack/generated/pkg/testcommon/ensure.go]: I made a note to change Ensure to something else, maybe `Verify`? Will do it when I have some spare time.
